units/en/unit3/agentic-rag/invitees.mdx

```diff
@@ -0,0 +1,175 @@
+
+# Load the dataset
+guest_dataset = datasets.load_dataset("agents-course/unit3-invitees", split="train")
+
+# Convert dataset entries into Document objects
+docs = [
+    Document(
+        page_content=f"Name: {guest['name']}\nRelation: {guest['relation']}\nDescription: {guest['description']}\nEmail: {guest['email']}",
+        metadata={"name": guest["name"]}
+    )
+    for guest in guest_dataset
+]
+
+```
+In the code above, we:
+- Load the dataset
+- Convert each guest entry into a `Document` object with formatted content
+- Store the `Document` objects in a list
+
+### Step 2: Create the Retriever Tool
+
+Now, let's create a custom tool that Alfred can use to search through our guest information.
+
+We will use the `BM25Retriever` from the `langchain_community.retrievers` module to create a retriever tool.
+
+<Tip>
+The `BM25Retriever` is a great starting point for retrieval, but for more advanced semantic search, you might consider using embedding-based retrievers like those from [sentence-transformers](https://www.sbert.net/).
+</Tip>
+
+```python
+from smolagents import Tool
+from langchain_community.retrievers import BM25Retriever
+
+class GuestInfoRetrieverTool(Tool):
+    name = "guest_info_retriever"
+    description = "Retrieves detailed information about gala guests based on their name or relation."
+    inputs = {
+        "query": {
+            "type": "string",
+            "description": "The name or relation of the guest you want information about."
+        }
+    }
+    output_type = "string"
+
+    def __init__(self, docs):
+        self.retriever = BM25Retriever.from_documents(docs)
+
+    def forward(self, query: str):
+        results = self.retriever.get_relevant_documents(query)
+        if results:
+            return "\n\n".join([doc.page_content for doc in results[:3]])
+        else:
+            return "No matching guest information found."
+
+# Initialize the tool
+guest_info_tool = GuestInfoRetrieverTool(docs)
+```
+
+Let's understand this tool. 
+- The `name` and `description` help the agent understand when and how to use this tool
+- The `inputs` define what parameters the tool expects (in this case, a search query)
+- We're using a `BM25Retriever`, which is a powerful text retrieval algorithm that doesn't require embeddings
+- The `forward` method processes the query and returns the most relevant guest information
+
+### Step 3: Integrate the Tool with Alfred
+
+Finally, let's bring everything together by creating our agent and equipping it with our custom tool:
+
+```python
+from smolagents import CodeAgent, HfApiModel
+
+# Initialize the Hugging Face model
+model = HfApiModel()
+
+# Create Alfred, our gala agent, with the guest info tool
+alfred = CodeAgent(tools=[guest_info_tool], model=model)
+
+# Example query Alfred might receive during the gala
+response = alfred.run("Tell me about our guest named 'Lady Ada Lovelace'.")
+
+print("🎩 Alfred's Response:")
+print(response)
+```
+
```

Comment:
````suggestion
Expected output:
```python
🎩 Alfred's Response:
Based on the information I retrieved, Lady Ada Lovelace is an esteemed mathematician and friend. She is renowned for her pioneering work in mathematics and computing, often celebrated as the first computer programmer due to her work on Charles Babbage's Analytical Engine. Her email address is ada.lovelace@example.com.
```
````